board.cpp

```diff
@@ -0,0 +1,190 @@
+}
+
+void Board::generateNewPiece() {
+	delete this->currentPiece;
+	this->currentPiece = this->pieceNames[rand() % PIECES::NUMBER_OF_PIECES]();
+	this->absPiecePositionCol = PIECES::INITIAL_ABS_POSITION_X;
+	this->absPiecePositionRow = PIECES::INITIAL_ABS_POSITION_Y;
+	this->rotation = PIECES::INITIAL_ROTATION;
+}
+
+void Board::generateNewPiece(Piece* piece) {
+	delete this->currentPiece;
+	this->currentPiece = piece;
+	this->absPiecePositionCol = PIECES::INITIAL_ABS_POSITION_X;
+	this->absPiecePositionRow = PIECES::INITIAL_ABS_POSITION_Y;
+	this->rotation = PIECES::INITIAL_ROTATION;
+
+}
+
+void Board::removeFilledRows(int rowStart) {
+	// Copy all unfilled rows from rowStart to currentHeight to an unfilled row
+	for (int currRow = rowStart, srcRow = rowStart; srcRow <= currentHeight;) {
+		if (this->blocksPerRow[srcRow] == BOARD::ROW_SIZE) {
+			++srcRow; // Skip this row as it is full and should not be in the result
+		}
+		else {
+			// copy srcRow to currRow
+			int currCell = currRow * BOARD::ROW_SIZE, currCellEnd = currRow + BOARD::ROW_SIZE,
+				srcCell = srcRow * BOARD::ROW_SIZE, srcCellEnd = srcCell + BOARD::ROW_SIZE;
+
+			for (; currCell < currCellEnd; ++currCell, ++srcCell) {
+				this->board[currCell] = this->board[srcCell];
+			}
+
+			++currRow, ++srcRow;
+		}
+	}
+}
+
+void Board::setPiece() {
+	this->currentHeight = std::max({ getPieceRow(0), getPieceRow(1), getPieceRow(2), getPieceRow(3) });
+
+	this->board[getPieceRowCol(0)] = this->currentPiece->color;
+	++this->blocksPerRow[getPieceRow(0)];
+
+	this->board[getPieceRowCol(1)] = this->currentPiece->color;
+	++this->blocksPerRow[getPieceRow(1)];
+
+	this->board[getPieceRowCol(2)] = this->currentPiece->color;
+	++this->blocksPerRow[getPieceRow(2)];
+
+	this->board[getPieceRowCol(3)] = this->currentPiece->color;
+	++this->blocksPerRow[getPieceRow(3)];
+}
+
+bool Board::movePieceDown() {
+	if (checkOutOfBoundsBelow(PIECES::MOVE_DOWN))
+		return false;
+
+	bool canFall = !checkPieceOverlaps(0, PIECES::MOVE_DOWN);
+
+	this->absPiecePositionRow += canFall * PIECES::MOVE_DOWN;
+
+	return canFall;
+}
+
+bool Board::movePieceLeft() {
+	if (checkOutOfBoundsLeft(PIECES::MOVE_LEFT)) {
+		return false;
+	}
+
+	bool canMove = !checkPieceOverlaps(PIECES::MOVE_LEFT, 0);
+
+	this->absPiecePositionCol += canMove * PIECES::MOVE_LEFT;
+	
+	return canMove;
+}
+
+bool Board::movePieceRight() {
+	if (checkOutOfBoundsRight(PIECES::MOVE_RIGHT))
+		return false;
+
+	bool canMove = !checkPieceOverlaps(PIECES::MOVE_RIGHT, 0);
+
+	this->absPiecePositionCol += canMove * PIECES::MOVE_RIGHT;
+
+	return canMove;
+};
+
+bool Board::rotatePiece() {
+	int oldRotation = this->rotation, oldPositionX = this->absPiecePositionCol, oldPositionY = this->absPiecePositionRow;
+	this->rotation = (this->rotation + 1) % PIECES::ROTATION_STATES;
```

Comment:
why not just update the rotation state once you've determined that its valid that seems not intuitive.

Reply:
The bounds checks use the member variables, they don't take the rotation or position as inputs so I save the old values, modify, check, and return them to normal if the piece can't move there.